### PR TITLE
Remove fluid:telemetry:ChannelDisposeError event, add error events where we fail to load/create data store

### PR DESCRIPTION
**Pull request by vladsud**
Please see issue #7802 for more details

### CODE REVIEW

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 1]
```diff
@@ -259,11 +259,7 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         if (this.channelDeferred) {
             this.channelDeferred.promise.then((runtime) => {
                 runtime.dispose();
-            }).catch((error) => {
-                this.logger.sendErrorEvent(
-                    { eventName: "ChannelDisposeError", fluidDataStoreId: this.id },
-                    error);
-            });
+            }).catch((error) => {});
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:262]: Add a comment explaining that we sent an error event in all codepaths where `channelDeferred` rejects?

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 2]
```diff
@@ -278,7 +274,10 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
             this.realizeCore(this.existing).catch((error) => {
                 this.channelDeferred?.reject(
                     CreateProcessingError(error, "realizeFluidDataStoreContext", undefined /* message */));
-            });
+                this.logger.sendErrorEvent(
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:277]: Couple things:
1. Let's log the same error we reject with
2. Should `this.id` be tagged as `packageData`? (Sorry I haven't renamed it yet btw)

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 3]
```diff
@@ -611,6 +610,9 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
             this.channelDeferred.resolve(this.channel);
         } catch (error) {
             this.channelDeferred?.reject(error);
+            this.logger.sendErrorEvent(
+                { eventName: "bindRuntimeError", fluidDataStoreId: this.id },
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:614]: Same comment as above.  Also, we should be consistent about casing for event names, at least within the same PR ;)  They seem to mostly be CamelCase, so probably stick with that?

**Review by markfields — COMMENTED** [review]
I think we should log the same error we reject with.  One other idea, which is probably over-engineering, is to make `channelDeferred` a light wrapper around `Deferred` that logs inside `reject`.  This way we don't introduce some other failure codepath that doesn't get logged? Not that I expect this code to change like that anytime soon.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:277]: I do not like # 1, as that may result in loss of information.
I'd rather see raw data, and developers can follow code to understand how data is transformed downstream

**Comment by markfields** [comment]
Trying to articulate the point of this change, looking again at #7802. Looks like this change will move the logging up to the moment the data store is being loaded/created and fails (as opposed to during cleanup), which will give us more info as to the state of the world at that point and how to interpret the errors?

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 2 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:278]: Look at what `prepareErrorObject` does and what `wrapError` does.  I think the only info preserved in either case is preserved in both:  `message` and `stack`.  I'll check for myself in a few min to make sure I'm right :)

**Comment by vladsud** [comment]
It will achieve two things:
1. Tell us exactly when user scenarios fail (possibly because of us, possible user error - we will keep clarifying it)
2. Do not report errors we do not care about (on disposal). Some of these errors might be not errors, as things are expected to fail when container runtime is disposed and some services are not working any more.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:278]: It does not matter what it preserves and what it does not (today). I prefer (where it's possible and does not make analyzes harder) to preserve original errors, as much as possible.
That said, if we want to bucket these errors together with other errors of similar type ("processing errors"), then yes, we might want to log wrapped error.

Comment (vladsud) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:278]: Ok, I changed it :)

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 4]
```diff
@@ -276,9 +273,12 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         if (!this.channelDeferred) {
             this.channelDeferred = new Deferred<IFluidDataStoreChannel>();
             this.realizeCore(this.existing).catch((error) => {
-                this.channelDeferred?.reject(
-                    CreateProcessingError(error, "realizeFluidDataStoreContext", undefined /* message */));
-            });
+                const errorWrapped = CreateProcessingError(error, "realizeFluidDataStoreContext");
+                this.channelDeferred?.reject(errorWrapped);
+                this.logger.sendErrorEvent(
+                    { eventName: "RealizeError", fluidDataStoreId: { value: this.id, tag: "PackageData"} },
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:279]: Why not put `fluidDataStoreId` prop on the error used for `reject` via `addTelemetryProps`?

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 5]
```diff
@@ -611,6 +611,9 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
             this.channelDeferred.resolve(this.channel);
         } catch (error) {
             this.channelDeferred?.reject(error);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:613]: Side note - Can you explain why this codepath doesn't use `CreateProcessingError`?

**Review by markfields — APPROVED** [review]
🎂

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 5 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:612]: DataProcessingError does not have comment describing when it should be used, but I'd guess the intention here is to track op processing errors, and that's not op processing error (most errors here are likely impacting local client only and are user errors).
That's my guess.

Comment (markfields) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:612]: Ok looking closer at this code, I guess this function is mostly runtime mechanics, plus pending op playback (any any errors in there will already be classified as DataProcessingError).  Not sure I see where user errors would arise, seems like mostly our own code. 🤷

**Pull request merged**